cron/worker/main.go

```
@@ -0,0 +1,136 @@
+	"github.com/ossf/scorecard/cron/pubsub"
+	"github.com/ossf/scorecard/pkg"
+	"github.com/ossf/scorecard/repos"
+	"github.com/ossf/scorecard/roundtripper"
+	"github.com/shurcooL/githubv4"
+	"go.uber.org/zap"
+)
+
+func processRequest(ctx context.Context,
+	batchRequest *data.ScorecardBatchRequest, bucketURL string,
+	httpClient *http.Client, githubClient *github.Client, graphClient *githubv4.Client) error {
+	repoURLs := make([]repos.RepoURL, len(batchRequest.GetRepos()))
+	var resultCh chan repos.RepoResult
+	var buffer bytes.Buffer
+	checksToRun := checks.AllChecks
+
+	for _, repo := range batchRequest.GetRepos() {
+		repoURL := repos.RepoURL{}
+		if err := repoURL.Set(repo); err != nil {
+			return fmt.Errorf("error setting RepoURL: %w", err)
+		}
+		if err := repoURL.ValidGitHubURL(); err != nil {
+			return fmt.Errorf("url is not a valid GitHub URL: %w", err)
+		}
+		repoURLs = append(repoURLs, repoURL)
+	}
+	go func() {
+		var wg sync.WaitGroup
+		for _, repoURL := range repoURLs {
+			repoURL := repoURL
+			wg.Add(1)
+			go func() {
+				resultCh <- pkg.RunScorecards(ctx, repoURL, checksToRun, httpClient, githubClient, graphClient)
+				wg.Done()
+			}()
+		}
+		wg.Wait()
+	}()
+
+	for result := range resultCh {
+		err := result.AsJSON(true /*showDetails*/, &buffer)
+		if err != nil {
+			return fmt.Errorf("error during result.AsJSON: %w", err)
+		}
+	}
+
+	filename := data.GetBlobFilename(
+		fmt.Sprintf("shard-%03d", batchRequest.GetShardNum()),
+		batchRequest.GetJobTime().AsTime())
+	return errors.Unwrap(data.WriteToBlobStore(ctx, bucketURL, filename, buffer.Bytes()))
+}
+
+func createNetClients(ctx context.Context) (httpClient *http.Client,
+	githubClient *github.Client, graphClient *githubv4.Client) {
+	cfg := zap.NewProductionConfig()
+	cfg.Level.SetLevel(zap.InfoLevel)
+	logger, err := cfg.Build()
+	if err != nil {
+		panic(err)
+	}
+	sugar := logger.Sugar()
+	// Use our custom roundtripper
+	rt := roundtripper.NewTransport(ctx, sugar)
```

I am guessing the `ENV` variables would be set in the deployment to the use the `HTTPCache`
```
 BLOB_URL:           gs://ossf-scorecards-cache
 USE_BLOB_CACHE:     true
 ```

---

Yes, the deployment of worker should be same/similar to cron job today. Let me know if anything is needed in code to ensure these vars are set.

---

How about validating if the ENV var is set and fail fast if it is not?

---

Isn't that validation already happening during the `roundtripper.NewTransport()` call?

---

No, It is optional. But in the pubsub worker, we need it.

---

Done. Added it in the `main` fn.